includes/derivatives.inc

```diff
@@ -0,0 +1,131 @@
+<?php
+
+/**
+ * @file
+ * Holds derivative functionality for compound solution pack.
+ */
+
+/**
+ * Derivative creation function for compound parent object TN.
+ *
+ * Wraps the TN update/creation function and
+ * adds expected logging.
+ *
+ * @param AbstractObject $object
+ *   The compound object.
+ * @param bool $force
+ *   Boolean for forcing this derivative creation.
+ *
+ * @return array
+ *   Success message.
+ */
+function islandora_compound_object_create_thumbnail(AbstractObject $object, $force = TRUE) {
+  if ($force || !isset($object['TN'])) {
+    $parts = islandora_compound_object_get_parts($object->id);
+    if (!empty($parts)) {
+      $child = islandora_object_load($parts[0]);
+      islandora_compound_object_update_parent_thumbnail($object, $child);
+    }
+    if (isset($child) && $object['TN'] == $child['TN']) {
+      return array(
+        'success' => TRUE,
+        'messages' => array(
+          array(
+            'message' => t('Updated @dsid on @parent.'),
+            'message_sub' => array(
+              '@dsid' => 'TN',
+              '@parent' => $object->id,
+            ),
+            'type' => 'dsm',
+            'severity' => 'status',
+          ),
+        ),
+      );
+    }
+    else {
+      return array(
+        'success' => FALSE,
+        'messages' => array(
+          array(
+            'message' => t('Failed to update @dsid on @parent.'),
+            'message_sub' => array(
+              '@dsid' => 'TN',
+              '@parent' => $object->id,
+            ),
+            'type' => 'watchdog',
+            'severity' => WATCHDOG_ERROR,
+          ),
+        ),
+      );
+    }
+  }
+}
+
+/**
+ * Update parent from child object.
+ *
+ * Given a child object, check to see if it is part of
+ * a compound and update the parent's TN datastream.
+ *
+ * @param AbstractObject $child
+ *   The potential child object.
+ * @param bool $force
+ *   Boolean for forcing this derivative creation.
+ *
+ * @return array|NULL
+ *   Success message.
+ */
+function islandora_compound_object_create_thumbnail_on_parent(AbstractObject $child, $force = TRUE) {
+  // Test if child has a parent.
+  $compound_info = islandora_compound_object_retrieve_compound_info($child);
+  if (empty($compound_info)) {
+    return NULL;
+  }
+  if ($force || !isset($compound_info['TN'])) {
+    // Test if child is the first child of the compound.
+    if (empty($compound_info['previous_pid'])) {
+      $parent = islandora_object_load($compound_info['parent_pid']);
+      // Child doesn't have a thumbnail then we can't set the parents TN either.
+      if (empty($child['TN'])) {
+        // Parents with first children with no TN shouldn't have one either.
+        if (isset($parent['TN'])) {
+          $parent->purgeDatastream('TN');
+        }
+        return NULL;
+      }
+      islandora_compound_object_update_parent_thumbnail($parent, $child);
+      if (isset($parent['TN']) && $parent['TN'] == $child['TN']) {
```

Comment:
The comparison of `TN` here appears to be making a mess:
```
Fatal error: Nesting level too deep - recursive dependency? in /var/www/drupal7/sites/all/modules/islandora_solution_pack_compound/includes/derivatives.inc on line 97 Call Stack: 0.0000 243744 
1. {main}() /var/www/drupal7/index.php:0 0.1089 4718696 
2. menu_execute_active_handler() /var/www/drupal7/index.php:21 0.1089 4720432 
3. call_user_func_array:{/var/www/drupal7/includes/menu.inc:527}() /var/www/drupal7/includes/menu.inc:527 0.1089 4720896 
4. drupal_get_form() /var/www/drupal7/includes/menu.inc:527 0.1089 4721952 
5. drupal_build_form() /var/www/drupal7/includes/form.inc:131 0.1101 4792720 
6. drupal_process_form() /var/www/drupal7/includes/form.inc:386 0.3244 6948416 
7. form_execute_handlers() /var/www/drupal7/includes/form.inc:904 0.3244 6952024 
8. islandora_datastream_version_replace_form_submit() /var/www/drupal7/includes/form.inc:1520 0.3249 6954816 
9. FedoraDatastream->setContentFromFile() /var/www/drupal7/sites/all/modules/islandora/includes/datastream.version.inc:343 0.3270 6961552 
10. IslandoraFedoraDatastream->modifyDatastream() /var/www/drupal7/sites/all/libraries/tuque/Datastream.php:1664 0.3731 7008064 
11. islandora_invoke_datastream_hooks() /var/www/drupal7/sites/all/modules/islandora/includes/tuque_wrapper.inc:598 0.3731 7009536 
12. islandora_invoke_hook_list() /var/www/drupal7/sites/all/modules/islandora/includes/tuque_wrapper.inc:75 0.3731 7011080 
13. call_user_func_array:{/var/www/drupal7/sites/all/modules/islandora/includes/utilities.inc:196}() /var/www/drupal7/sites/all/modules/islandora/includes/utilities.inc:196 0.3731 7011696 
14. module_invoke_all() /var/www/drupal7/sites/all/modules/islandora/includes/utilities.inc:196 0.3733 7014864 
15. call_user_func_array:{/var/www/drupal7/includes/module.inc:957}() /var/www/drupal7/includes/module.inc:957 0.3733 7015464 
16. islandora_islandora_datastream_modified() /var/www/drupal7/includes/module.inc:957 0.3734 7017832 
17. islandora_do_derivatives() /var/www/drupal7/sites/all/modules/islandora/islandora.module:1804 0.3786 7026616 
18. call_user_func:{/var/www/drupal7/sites/all/modules/islandora/includes/derivatives.inc:93}() /var/www/drupal7/sites/all/modules/islandora/includes/derivatives.inc:93 0.3786 7026776 
19. islandora_basic_image_create_thumbnail() /var/www/drupal7/sites/all/modules/islandora/includes/derivatives.inc:93 0.4462 7035992 
20. islandora_basic_image_add_datastream() /var/www/drupal7/sites/all/modules/islandora_solution_pack_image/includes/derivatives.inc:115 0.4463 7036416 
21. FedoraDatastream->setContentFromFile() /var/www/drupal7/sites/all/modules/islandora_solution_pack_image/includes/derivatives.inc:75 0.4497 7046696 
22. IslandoraFedoraDatastream->modifyDatastream() /var/www/drupal7/sites/all/libraries/tuque/Datastream.php:1664 0.5003 7052976 
23. islandora_invoke_datastream_hooks() /var/www/drupal7/sites/all/modules/islandora/includes/tuque_wrapper.inc:598 0.5003 7054448 
24. islandora_invoke_hook_list() /var/www/drupal7/sites/all/modules/islandora/includes/tuque_wrapper.inc:75 0.5003 7055984 
25. call_user_func_array:{/var/www/drupal7/sites/all/modules/islandora/includes/utilities.inc:196}() /var/www/drupal7/sites/all/modules/islandora/includes/utilities.inc:196 0.5003 7056600 
26. module_invoke_all() /var/www/drupal7/sites/all/modules/islandora/includes/utilities.inc:196 0.5003 7058688 
27. call_user_func_array:{/var/www/drupal7/includes/module.inc:957}() /var/www/drupal7/includes/module.inc:957 0.5003 7059208 
28. islandora_islandora_datastream_modified() /var/www/drupal7/includes/module.inc:957 0.5004 7060128 
29. islandora_do_derivatives() /var/www/drupal7/sites/all/modules/islandora/islandora.module:1804 0.5059 7066528 
30. call_user_func:{/var/www/drupal7/sites/all/modules/islandora/includes/derivatives.inc:93}() /var/www/drupal7/sites/all/modules/islandora/includes/derivatives.inc:93 0.5059 7066680 
31. islandora_compound_object_create_thumbnail_on_parent() /var/www/drupal7/sites/all/modules/islandora/includes/derivatives.inc:93 
```

... [The non-strict equivalence check between objects compares all properties of said objects](http://www.php.net/manual/en/language.oop5.object-comparison.php), which it would seems gets into recursive looping.

Seems like it could be expensive, but given they're thumbnails, they shouldn't be too big... Could do something like:
```php
$parent['TN']->content === $child['TN']->content
```